Add DataFrameTransformer

BenjaminBossan opened the pull request:
Addresses #501

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,268 @@ def predefined_split(dataset):
+        for col, dtype in zip(df, df.dtypes):
+            if isinstance(dtype, self.pd.core.dtypes.dtypes.CategoricalDtype):
+                continue
+            if dtype in self.int_dtypes_:
```
Does `np.issubdtype(dtype, np.integer)` work?

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,268 @@ def predefined_split(dataset):
+                continue
+            if dtype in self.int_dtypes_:
+                continue
+            if dtype in self.float_dtypes_:
```
Does `np.issubtype(dtype, np.floating)` work?

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,268 @@ def predefined_split(dataset):
+      values will have different dtypes, reflecting the number of
+      unique categories.
+
+    signature_ : namedtuple
```
A named tuple defined inside a class seems a little harder to discover. Would at tuple or a dictionary be enough?

BenjaminBossan commented:
I changed the code to use `np.issubdtype` (learned something new). Moreover, I dropped the namedtuple but instead of using a tuple, I used a dict. The reason for that is the better repr, otherwise a lonely number is not very easy to interpret.

@thomasjpfan @ottonemo pls review again

BenjaminBossan commented:
@thomasjpfan @ottonemo any things to add to this?

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
+    ...         super().__init__()
+    ...         self.reset_params()
+
+    >>>     def reset_params(self):
```
Should this be directly in `__init__`?

BenjaminBossan replied:
It's just my personal preference, that way it's possible to reset the module without creating a new instance.

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
+        wrong_dtypes = []
+
+        for col, dtype in zip(df, df.dtypes):
+            if isinstance(dtype, self.pd.core.dtypes.dtypes.CategoricalDtype):
```
Does `pd.api.types.CategoricalDtype` work?

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
+        Xf = []  # floats
+
+        for col, dtype in zip(df, df.dtypes):
+            values = df[col]
```
Nit: Having `values.values` later is a bit harder to follow. Maybe `X_col=df[col`?

BenjaminBossan replied:
Not a nit :+1:

BenjaminBossan commented:
@ottonemo is this ready to be merged?

thomasjpfan commented on skorch/tests/test_helper.py:
```diff
@@ -474,3 +482,230 @@ def test_pickle(self, predefined_split, data):
+            'col_floats': [0.1, 0.2, 0.3],
+            'col_ints': [11, 11, 10],
+            'col_cats': ['a', 'b', 'a'],
+        })
```
I am a little surprised some of the test pass without this in python 3.5:

```suggestion
        }, columns=['col_floats', 'col_ints', 'col_cats'])
```

BenjaminBossan replied:
Why indicate the column names a second time?

thomasjpfan commented on skorch/helper.py:
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
+                continue
+
+            if (
+                    dtype in (np.dtype('int32'), np.dtype('int64'))
```
```suggestion
                    np.issubdtype(dtype, np.integer):
```

ottonemo commented:
I will merge this after the tests complete since this seems to be a version we can work with / build up on. What I would like to see in the future is a way to deal with NULL-values for categorical columns (introduce a separate category for NULL) since this is a common enough problem.

thomasjpfan commented:
Since this PR uses the `cat.codes`, the `np.nan` values will be encoded as `-1`:

```py
import pandas as pd
df = pd.DataFrame({
    'col': pd.Categorical(['a', 'b', np.nan, 'b'])})

df['col'].cat.codes.values
# array([ 0,  1, -1,  1], dtype=int8)
```
